### ISSUE TITLE
mbp: Provide state accessors and mutators from MultibodyPlant; make them friendly and value-caching based

### ISSUE THREAD

**Issue by EricCousineau-TRI** [RobotLocomotion/drake#9762]
At present, to set MBP context, you may have to jump through a few hoops, such as finding the correct subsystem context (`diagram.GetSubsystemContext(...)`), casting it to the correct type (`dynamic_cast<MultibodyTreeContext*>(...)`), and then mutating the values. Even then, since the Systems framework caching is ticket-based, if you scribble the same numbers twice, you'll invalidate your cache.

For MBP, it'd be nice to provide sugar accessors / mutators which:
(a) Are vector-valued and simple
(b) [maybe / mebbe not] Don't care if they have a parent system's context. They'll find the right thing.
(c) [maybe / mebbe not] Do value-based caching, as in [IK](https://github.com/RobotLocomotion/drake/blob/d2a2087/multibody/inverse_kinematics/kinematic_constraint_utilities.cc#L23-L28).

There may be some performance penalties to start, but I'm sure we can work through those.

@sherm1 Is there an easy way to achieve (b) given the current systems framework, without MBP having to be aware of it's parent diagram?

TL;DR: Move / duplicate accessors / mutators from `MultibodyTreeContext` to `MultibodyPlant`.

\cc @hongkai-dai @sherm1

**Comment by jwnimmer-tri** [comment]
I don't think (b) is important.  We can grab the right subcontext once, at construction time of the using-code, and never worry about it again.

**Comment by sherm1** [comment]
I'm definitely :+1: for getting rid of MultibodyTreeContext (there's a TODO already for me).

I'm not sure MBPlant should have unusual context superpowers as you propose in (b) though, even if that could be implemented.

**Comment by sherm1** [comment]
P.S. Value-based caching is _not_ an unambiguous win. If the expectation is that values have changed in nearly every set() call, or if the underlying computation is cheap, it is a loss. In simulation we rarely expect to see the same value set twice. In optimization it may be more common. I would not advocate putting it in MBP without some performance studies on meaningful use cases.

**Comment by EricCousineau-TRI** [comment]
Given that items (b) and (c) are mebbe nots for now, will set this to be resolved via #9977.
